### ISSUE TITLE
Experimental API client should be part of the client.Interface

### ISSUE THREAD

**Issue by rrati** [kubernetes/kubernetes#13388]
Currently the experimental API client is a separate entity from the stable client interface.  This causes problems if anything in the experimental API needs to be accessed by any component of kubernetes.  The only way a kubernetes process can access anything in the experimental API is to change the function signatures for anything that needs access to the experimental API.  This has large consequences for anything using/launching that component as well as any test case that uses that component.

An easy example here is anything that is namespaced.  If you add a new entity that is namespaced, you need to modify the namespace controller to accept an experimental client which impacts all sorts of other bits that invoke a namespace controller, like the mesos integration and tests.

**Comment by rrati** [comment]
@lavalamp @bgrant0607 @smarterclayton @piosz @timstclair @derekwaynecarr

**Comment by piosz** [comment]
cc @mwielgus @jszczepkowski

**Comment by derekwaynecarr** [comment]
My initial thought is that the unversioned client should support the ability to reference an experiemental client via an accessor.

**Comment by derekwaynecarr** [comment]
Or we should pass around a client library instead of client.Interface directly and then clients can get the client object they need by version/type.  Agree we definitely need a semantic for this to keep things sane.  Passing client.Interface around now is not sufficient.

**Comment by timothysc** [comment]
It seems fairly trivial to add to client.Interface, but we just want folks to agree on this one.

**Comment by mikedanese** [comment]
We should to decide one way or the other for _all_ future API groups. Experimental should not be treated separately from those. @caesarxuchao is the one working on this. Please comment on #12951 (API Group RFC) which has a section on client-side implementation.

**Comment by mikedanese** [comment]
We cannot assume that the experimental API is always turned on. We should also be weary of experimental polluting stable systems.

> An easy example here is anything that is namespaced. If you add a new entity that is namespaced, you need to modify the namespace controller to accept an experimental client which impacts all sorts of other bits that invoke a namespace controller

Is the namespace controller an example or an exception? If an existing and stable controller suddenly needs to depend on an experimental API, that seems broken. Do we need to use the experimental client for other existing controllers?

**Comment by timothysc** [comment]
@mikedanese it's an example of how it breaks down.  Other users of the client.Interface will need access to experimental api.

**Comment by mikedanese** [comment]
@timothysc I buy it. Then the burden is on the reviewers to make sure experimental doesn't leak. FWIW, I hackily added Experimental to client.Interface in https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd in #13182 to get e2e tests to pass until this is solved. I was considering the Daemon Controller PR (and now #13037?) to be blocked until API Group RFC (#12951) is merged. Maybe this will motivate people to speed up the review. :)

**Comment by lavalamp** [comment]
We need to add a layer to our api. What is now `c.Pods("foo").Create...` needs to become `c.Core().Pods("foo").Create...`. (Of course, we can pass around c.Core() instead of just c, to reduce code churn.)

The experimental group should be accesed as `c.Experimental().Foo("ns").Create...`.

We could consider moving the namespace parameter into the api group call, but that'd be a much larger change.

**Comment by timothysc** [comment]
@mikedanese yes please, we are blocked on this. 

@lavalamp - I LIKE IT! So when can we get it?    ;-)

**Comment by derekwaynecarr** [comment]
@timothysc  @lavalamp - i am fine poking through the client patterns to provide a grouping above client.Interface while I wait on some other items to land, but I would want some agreement on naming before proceeding.  I had been burned going too far on client changes in the past before we settled on the *Namespacer pattern ;-)  So if someone wants to put a quick proposal in here that is acceptable I am fine to hack on it.

**Comment by timothysc** [comment]
@lavalamp aren't you the gatekeeper here? 

> I had been burned going too far on client changes in the past before we settled on the ...

@derekwaynecarr +10e256.

**Comment by caesarxuchao** [comment]
I already have some work towards this direction on my local branch, but it depends on #13233. 

I will incorporate lavalamp's comment https://github.com/kubernetes/kubernetes/issues/13388#issuecomment-136427016 in #12591.

**Comment by lavalamp** [comment]
I will happily take a change that adds a layer in front. c.Core(), c.Experimental(), etc.

**Comment by caesarxuchao** [comment]
@derekparke could you wait for my PR which depends on #13233?

**Comment by derekwaynecarr** [comment]
sure

**Comment by timothysc** [comment]
@caesarxuchao what's your ETA, b/c if it's a couple of days out I'm just going to implement.

**Comment by caesarxuchao** [comment]
@timothysc, if you urgently needs the PR, could you take a look at https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd? Basically ot implements the c.Experimental(). Our main concern is how it makes the experimentalConfig [here](https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd#diff-f013b1aafa99ed0c479d939400d636e9R142).

cc @mikedanese

**Comment by mikedanese** [comment]
@timothysc I went down that path and found that to get client.Config and version negotiation working properly in the general case is pretty involved. I recommend doing as little as possible to get unblocked and e2e passing, then you can rebase over @caesarxuchao's work. The client.Interface interface will likely be the same or only slightly different so rebase should be straightforward.

**Comment by caesarxuchao** [comment]
@timothysc, after in-person discussion with mikedanese, we think you can incorporate https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd to unblock your PR. We can deal with the experimentalConfig in a separate PR.

**Comment by lavalamp** [comment]
I agree that https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd looks like a good start.

**Comment by bgrant0607** [comment]
deleteAllContent shouldn't really hardcode all the namespaced resources. That's not at all extensible. (And it's not documented in api_changes.md, AFAIK.) Like kubectl, it should be generic.

The API discovery mechanism(s) we're discussing as part of API groups would improve this greatly.

**Comment by timothysc** [comment]
The original implementation pretty much needs to get ripped out.

**Comment by uluyol** [comment]
Please see https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37546133. I think we should avoid having a "kubernetes client" which is distinct from an "openshift client" (in terms of type definitions). We could have pkg/kubectl/resource.Helper's wrapping RESTClients for every group, and this would be completely generic.

A small change I would make (as mentioned in the discussion), is that I would change the functions that return runtime.Objects to take one as a parameter. It would reduce the amount of error handling and type conversions.

**Comment by lavalamp** [comment]
> A small change I would make (as mentioned in the discussion), is that I would change the functions that return runtime.Objects to take one as a parameter.

Typed client methods should not return runtime.Object, they should return the concrete type (e.g., Pod). See: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/pods.go#L85

**Comment by caesarxuchao** [comment]
@timothysc, why do we need to undo #12405?  https://github.com/mikedanese/kubernetes/commit/836a473a07ad3009c8a6253df07ca7d16f4951bd  works with #12405.

If you are going to do major changes I'd rather wait until the multiple group support (#12888) is merged.

**Comment by caesarxuchao** [comment]
@timothysc, I guess you meant to removing ExperimentalClient() and related stuff from #12405. I'm fine with that.

**Comment by uluyol** [comment]
@lavalamp, should we continue to have typed client methods? How does this work with multiple api groups that are versioned independently? I think that running the following isn't too different from our current client.
```
var pod *api.Pod
helper := resource.NewHelper(restClient, latest.RESTMapper.RESTMapping("Pod", "v1'))
podObj, err := helper.Get("default", "mypod") // passing the pod in here would remove the type assertion below
...
pod, ok = podObj.(*Pod)
```
We could probably clean up resource.Helper so it's simpler to use. So instead of dealing with a Client (and implicitly a RESTClient), we use a resource.Helper and RESTClient.

**Comment by lavalamp** [comment]
Typed clients are nice, it lets the compiler help you. We should autogenerate them.

Generic clients are good too, and those should use runtime.Objects. 

We want both.

**Comment by timothysc** [comment]
@caesarxuchao Namely referencing the kubectl work.

**Comment by uluyol** [comment]
@lavalamp, agreed, but how will we handle versioned clients? With v2 onwards won't we have multiple groups that evolve at different rates? Will we just pick a set of versions for those groups and declare it "the set kubernetes API versions" for vN?

**Comment by caesarxuchao** [comment]
I think we should allow user to express which version of a group to use in the client config file. Also, all groups should support the major versions vN. If the user picks vN for the entire client, then all typed clients should use vN in default.

**Comment by lavalamp** [comment]
@uluyol 
> but how will we handle versioned clients?

Versioned clients will be compiled against a specific set of api types. It will take a little bit of thought to do this right, thanks for mentioning it. Worst case, we can offer things like `c.ExperimentalAlpha1().Deployment()...`.

**Comment by lavalamp** [comment]
@caesarxuchao 

> If the user picks vN for the entire client, then all typed clients should use vN in default.

Versions across groups are not logically related (the apis of different groups could evolve at different rates), so we shouldn't do this.

**Comment by timothysc** [comment]
At this point I don't want to conflate the changes, or wait on other potentially destabilizing mods. 

I would prefer just to fix experimental, and look at the other issue in their own right.

**Comment by rrati** [comment]
Fixed by #13523
